### DISCUSSION TITLE
ACL not applying on deployment update

### DISCUSSION THREAD

**Discussion by MacherelR** [emqx/emqx-operator#972] [Q&A]
Hi everyone, 
I've been trying for a moment now to create and apply ACLs to my emqx cluster. 
My system is the following: Deployed emqx-operator following the getting started from Readme.md, and initially applied basic emqx deployment. 
Now my deployment.yaml file looks as follows:
~~~
apiVersion: v1
kind: ConfigMap
metadata:
  name: emqx-acl-config
data:
  acl.conf: |
    {allow, all, subscribe, ["registration/request"]}.
    {deny, all, subscribe, ["$SYS/#", "#"]}.
    {allow, {user, {re, "^device-[0-9a-zA-Z]+$"}}, subscribe, ["private/%u/#"]}.
    {allow, {user, {re, "^device-[0-9a-zA-Z]+$"}}, publish, ["private/%u/#"]}.
---
apiVersion: apps.emqx.io/v2beta1
kind: EMQX
metadata:
   name: emqx
spec:
   image: emqx:5.1
   imagePullPolicy: IfNotPresent
   config:
      data: |
         listeners.tcp.test {
            bind = "0.0.0.0:1884"
            max_connections = 1024000
         }
         allow_anonymous = false
         acl_nomatch = deny
   coreTemplate:
      spec:
         extraVolumes:
           - name: my-owner-acl
             configMap:
               name: emqx-acl-config
         extraVolumeMounts:
           - name: my-owner-acl
             mountPath: /opt/emqx/etc/acl.conf
             subPath: acl.conf
   replicantTemplate:
      spec:
         extraVolumes:
           - name: my-owner-acl
             configMap:
               name: emqx-acl-config
         extraVolumeMounts:
           - name: my-owner-acl
             mountPath: /opt/emqx/etc/acl.conf
             subPath: acl.conf
   listenersServiceTemplate:
      spec:
         type: LoadBalancer
   dashboardServiceTemplate:
      spec:
         type: LoadBalancer 
~~~
As you can see, I wrote the configmap declaration in the same file so that only one kubectl apply command is needed.
But when I do apply this file I get configmap/emqx-acl-config configured and emqx.apps.emqx.io/emqx configured 
When this is the first deployment everything works like charm but when I modify my ACLs and redo the kubectl apply -f command the acls don't seem tu update themselves, is there a way to force restart the deployment or the pods ? 

Rémy

**Comment by Rory-Z** [comment]
Hi @MacherelR I'm sorry, EMQX Operator can not auto reloader ACL file, I think you can reload ACL file by two ways:
1. You can restart EMQX pods after update ACL file .
2. You can disable and enable EMQX Authorization by EMQX APIs, please check this document: https://docs.emqx.com/en/enterprise/v5.3/admin/api-docs.html#tag/Authorization/paths/~1authorization~1sources~1%7Btype%7D/put

**Reply by MacherelR** [reply]
Hello @Rory-Z, So I've now turned into JWT authentication as it seemed to fit my use case better. Only blurred point in it is that I can only see documentation on how to configure jwt from the dashboard.. ([Here](https://www.emqx.io/docs/en/latest/dashboard/authn.html)) 
I figured that it could be done also in the config section of the deployment file with something such as :
~~~
config:
      data: |
         listeners.tcp.test {
            bind = "0.0.0.0:1884"
            max_connections = 1024000
         }
         authentication = [{
            from = password
            mechanism = jwt
            enable = true
            secret = emqxsecret
            secret_base64_encoded = false
            algorithm = hmac-based
            use_jwks = false
            verify_claims {
               clientid = "${clientid}"
            }
            acl_claim_name = "acl"
         }]
~~~
Am I right with this ? If yes, I'm not quite sure on what should be the secret used (how to define it) and then should the token be something like : 
~~~
{
  "username": "test-user",
  "clientid": "mqttx_4f52cd40",
  "acl": {
    "pub": [
      "test/user/",
      "test/device"
    ],
    "sub": [
      "test/device"
    ]
  }
}
~~~
And if I understood well this means the token should be encoded using the secret ? Or am I completely wrong on this one ? I'm sorry but I don't have lots of knowldege regarding to security so...

**Reply by Rory-Z** [reply]
Hello @savonarola could you please take a look

**Reply by savonarola** [reply]
Hello!

You should use secret to sign the token. Example for jwt.io:

<img width="1123" alt="20231122121549-qltdppq1qer69flvs2a8" src="https://github.com/emqx/emqx-operator/assets/4236/adaa071f-6a25-49c4-ad6b-41517db617ea">


Then, a client will be able to connect, passing
* the generated token as password
* `mqttx_4f52cd40` as clientid
* arbitrary string username

**Reply by savonarola** [reply]
Aslo please note that specifying secret in k8s config is not secure.

Maybe you should consider taking it from a k8s secret and passing it to EMQX through env.

**Reply by MacherelR** [reply]
Okay thank you very much !
So if I resume I should define a secret for my namespace in k8s such as for example :
~~~
apiVersion: v1
kind: Secret
metadata:
  name: emqx-secret
type: Opaque
data:
  emqxsecret: <SECRETVal>
~~~
Then generate (for testing purposes) a jwt token in jwt.io by specifying SECRETVal in the field you showed me.
But now how should I link it to emqx conf value ? 
should I do something like :
~~~ 
authentication = [{
            from = password
            mechanism = jwt
            enable = true
            secret = EMQXSECRET (how to specify it is an env variable ?)
            secret_base64_encoded = true
            algorithm = hmac-based
            use_jwks = false
            verify_claims {
               clientid = "${clientid}"
            }
            acl_claim_name = "acl"
         }]
~~~
And then define an env Variable with the SECRETVal for the emqx deployment is that right ?
Thanks a lot in advance for your help and expertise !

**Reply by savonarola** [reply]
Hi! You may:
* just skip secret key in the config part
* mount the secret as `EMQX_AUTHENTICATION__1__SECRET` env variable to the pods

EMQX will merge config and envs.
